File: backend/app/domain/entities/plan.py
"""The training plan: what a trainer schedules for a student, day by day.

A plan item is a *prescription*. Whether it happened is not stored here — it is
read from the student's workout log for that exercise and day, so there is one
source of truth for "what was lifted" and the dashboard charts pick up assigned
work automatically.
"""

from dataclasses import dataclass
from datetime import date
from enum import StrEnum


class PlanItemStatus(StrEnum):
    """How a scheduled exercise stands today."""

    #: Scheduled for today or later, nothing logged yet.
    PENDING = "pending"
    #: Logged, hitting the target.
    DONE = "done"
    #: Logged, but short of the target weight or reps.
    PARTIAL = "partial"
    #: The day passed with nothing logged.
    MISSED = "missed"


@dataclass(frozen=True)
class PlanItem:
    """One exercise scheduled for one day, with whatever the student logged."""

    id: int
    trainer_id: int
    student_id: int
    exercise_id: int
    exercise_name: str
    scheduled_on: date
    target_sets: int
    target_reps: int
    #: None means "no target weight" — bodyweight work, or the trainer left it open.
    target_weight_kg: float | None
    notes: str | None
    #: What the student actually logged that day, if anything.
    done_weight_kg: float | None = None
    done_reps: int | None = None
    done_sets: int | None = None
    done_completed: bool | None = None

    @property
    def is_logged(self) -> bool:
        return self.done_completed is not None

    def status(self, today: date) -> PlanItemStatus:
        """Resolve the status against the calendar."""
        if not self.is_logged:
            return PlanItemStatus.MISSED if self.scheduled_on < today else PlanItemStatus.PENDING
        if not self.done_completed:
            return PlanItemStatus.PARTIAL
        # Logged as completed, but under the prescribed load is still partial:
        # the student did the work, not the work that was asked for.
        short_on_weight = (
            self.target_weight_kg is not None
            and self.done_weight_kg is not None
            and self.done_weight_kg + 0.01 < self.target_weight_kg
        )
        short_on_reps = self.done_reps is not None and self.done_reps < self.target_reps
        short_on_sets = self.done_sets is not None and self.done_sets < self.target_sets
        short = short_on_weight or short_on_reps or short_on_sets
        return PlanItemStatus.PARTIAL if short else PlanItemStatus.DONE

File: frontend/src/api/plan.ts
// Training calendar: what the trainer schedules and what the student reports.

import { api } from './client'
import { i18n } from '@/i18n'

export type PlanItemStatus = 'pending' | 'done' | 'partial' | 'missed'

export interface PlanItem {
  id: number
  exerciseId: number
  exerciseName: string
  scheduledOn: string // YYYY-MM-DD
  targetSets: number
  targetReps: number
  targetWeightKg: number | null
  notes: string | null
  doneWeightKg: number | null
  doneReps: number | null
  status: PlanItemStatus
}

export interface ScheduleRequest {
  exerciseId: number
  scheduledOn: string
  targetSets?: number
  targetReps?: number
  targetWeightKg?: number | null
  notes?: string | null
}

interface PlanItemPayload {
  id: number
  exercise_id: number
  exercise_name: string
  scheduled_on: string
  target_sets: number
  target_reps: number
  target_weight_kg: number | null
  notes: string | null
  done_weight_kg: number | null
  done_reps: number | null
  status: PlanItemStatus
}

function toItem(payload: PlanItemPayload): PlanItem {
  return {
    id: payload.id,
    exerciseId: payload.exercise_id,
    exerciseName: payload.exercise_name,
    scheduledOn: payload.scheduled_on,
    targetSets: payload.target_sets,
    targetReps: payload.target_reps,
    targetWeightKg: payload.target_weight_kg,
    notes: payload.notes,
    doneWeightKg: payload.done_weight_kg,
    doneReps: payload.done_reps,
    status: payload.status,
  }
}

function range(from: string, to: string): string {
  return `from=${from}&to=${to}&lang=${i18n.global.locale.value}`
}

/** The signed-in student's own calendar. */
export async function myPlan(from: string, to: string): Promise<PlanItem[]> {
  const payload = await api.get<PlanItemPayload[]>(`/coaching/me/plan?${range(from, to)}`)
  return payload.map(toItem)
}

/** The calendar of one of the trainer's students. */
export async function studentPlan(
  studentId: number,
  from: string,
  to: string,
): Promise<PlanItem[]> {
  const payload = await api.get<PlanItemPayload[]>(
    `/coaching/students/${studentId}/plan?${range(from, to)}`,
  )
  return payload.map(toItem)
}

/** Schedule an exercise, or edit the targets already set for that day. */
export async function scheduleExercise(
  studentId: number,
  request: ScheduleRequest,
): Promise<PlanItem> {
  const payload = await api.post<PlanItemPayload>(
    `/coaching/students/${studentId}/plan?lang=${i18n.global.locale.value}`,
    {
      exercise_id: request.exerciseId,
      scheduled_on: request.scheduledOn,
      target_sets: request.targetSets,
      target_reps: request.targetReps,
      target_weight_kg: request.targetWeightKg,
      notes: request.notes,
    },
  )
  return toItem(payload)
}

export function unscheduleExercise(itemId: number): Promise<void> {
  return api.remove(`/coaching/plan/${itemId}`)
}

/** Report what was lifted; returns the item with its status recomputed. */
export async function reportPlanItem(
  itemId: number,
  weightKg: number,
  reps: number,
  completed: boolean,
): Promise<PlanItem> {
  const payload = await api.post<PlanItemPayload>(
    `/coaching/me/plan/${itemId}/report?lang=${i18n.global.locale.value}`,
    { weight_kg: weightKg, reps, completed },
  )
  return toItem(payload)
}
